### PR TITLE
Database Creation

### PR DESCRIPTION
**Pull request by johguentner**
### Adding multiple ways to create a new database within the Notion API

```php
$scheme = PropertyBuilder::bulk()
                ->title('MyTitle') // required
                ->richText('MyText')
                ->checkbox('...')
                ->status()
                ->number()
                ->date()
                ->url()
                ->email()
                ->phoneNumber()
                ->people()
                ->files()
                ->createdBy()
                ->createdTime()
                ->lastEditedBy()
                ->lastEditedTime()
                ->formula('Testing', 'prop("MultiSelect")')
                ->select('Select', [
                    [
                        'name' => 'option 1 select',
                        'color' => 'blue'
                    ]
                ])->multiSelect('MultiSelect', [
                    [
                        'name' => 'option 1 multiselect',
                        'color' => 'yellow'
                    ]
                ]);

 $databaseEntity = Notion::databases()
                  ->build()
                  ->inline()
                  ->title('Inline Test DB 1')
                  ->add($scheme)
                  ->createInPage('1c682e7371ec4399be1cc015686c67c6');
```

### To keep it a bit more similar to Laravel migrations, it's also possible to do it like this:

```php

$databaseEntity = Notion::databases()
                  ->build()
                  ->title('Test DB 2')
                  ->scheme(function ($table){
                      $table->title('My Title');
                      $table->richText('My Text');
                      $table->date('My Date');
                  })
                  ->createInPage('1c682e7371ec4399be1cc015686c67c6');

```

### To keep things simple, it's also possible to use the `PropertyBuilder::class`.

```php
$props = collect([
     PropertyBuilder::title(),
     PropertyBuilder::richText(),
     ...
]);

$databaseEntity = Notion::databases()
                  ->build()
                  ->title('Test DB 3')
                  ->add($props) //add multiple props
                  ->add(PropertyBuilder::date('My Date')) //add single props
                  ->createInPage('1c682e7371ec4399be1cc015686c67c6');

```

### Polish of `PestHttpRecorder::class`

- remove adding the query/payload to the snapshot filename
- instead add short hash of query/payload to the snapshot filename
- or (if given) allow user to set a specific name for the upcoming request
- additionally writing header, method and payload to the snapshot

### CODE REVIEW

**Comment by farez** [comment]
Hello Johannes!

Ah that looks awesome. Should work well.

An idea: have a shortcut method called addProperties() that takes an array
of property definitions. Saves calling addProperty lots of times.

Thanks man. Look forward to using this.

Farez

**Comment by johguentner** [comment]
> Hello Johannes! Ah that looks awesome. Should work well. An idea: have a shortcut method called addProperties() that takes an array of property definitions. Saves calling addProperty lots of times. Thanks man. Look forward to using this. Farez

Thank you for your input! 
I updated the description above to show all possibilities. 
After I add the final touches + especially testing, I'll push this as `1.1.0`

File: src/Builder/DatabaseBuilder.php [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+     * @var array
+     */
+    private array $payload;
+
+    /**
+     * DatabaseBuilder constructor.
+     *
+     * @param  Databases  $databasesEndpoint
+     */
+    public function __construct(private Databases $databasesEndpoint)
```
Comment (mechelon) [inline src/Builder/DatabaseBuilder.php:24]: If we don't use `$databasesEndpoint` in the constructor, can we define it next to payload up there?

File: src/Builder/DatabaseBuilder.php [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        ];
+    }
+
+    /**
+     * Creates database within given page.
+     *
+     * @param  string  $pageId
+     * @return Database
+     */
+    public function createInPage($pageId): Database
```
Comment (mechelon) [inline src/Builder/DatabaseBuilder.php:46]: Can we type `$pageId`?

File: src/Builder/DatabaseBuilder.php [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        return $this;
+    }
+
+    /**
+     * Adds the property `title` database creation.
+     *
+     * @param  string  $name
+     * @return DatabaseBuilder
+     */
+    public function addTitle(string $name = 'Name')
```
Comment (mechelon) [inline src/Builder/DatabaseBuilder.php:170]: Missing return type? 🤔

File: src/Builder/DatabaseBuilder.php [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+    /**
+     * Adds a raw property to the database creation.
+     *
+     * @param  string  $title
+     * @param  string  $propertyType
+     * @param  array|null  $content
+     * @return DatabaseBuilder
+     */
+    public function addRaw(string $title, string $propertyType, array $content = null): DatabaseBuilder
```
Comment (mechelon) [inline src/Builder/DatabaseBuilder.php:222]: I think it should be `?array $content = null` right? 🤔

File: src/Endpoints/Resolve.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                return $this->notion->pages()->find($parent->getId());
+            case 'database_id':
+                return $this->notion->databases()->find($parent->getId());
+            case 'block_id':
+                return $this->notion->block($parent->getId())->retrieve();
+            case 'workspace':
+                return $parent;
+                // throw new HandlingException('A Notion Workspace cannot be resolved by the Notion API.');
+            default:
+                throw new HandlingException('Unknown parent type while resolving the notion parent');
```
Comment (mechelon) [inline src/Endpoints/Resolve.php:90]: notion -> Notion

File: src/Macros/PestHttpRecorder.php [hunk 6]
```diff
@@ -37,6 +38,8 @@ class HttpRecorder
 
     private $usePrettyJson = true;
 
+    private $requestNames = [];
```
[the remark this answers is not in the document]
Comment (johguentner) [reply src/Macros/PestHttpRecorder.php:41]: Yep, added them for all attributes

File: src/Entities/NotionParent.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        parent::setResponseData($responseData);
+
+        if (
+            $responseData['object'] !== 'page_id'
+            && $responseData['object'] !== 'database_id'
+            && $responseData['object'] !== 'workspace'
+            && $responseData['object'] !== 'block_id'
+        ) {
+            throw HandlingException::instance('invalid json-array: the given object is not a valid parent');
```
[the remark this answers is not in the document]
Comment (johguentner) [reply src/Entities/NotionParent.php:28]: yep, done :)

File: src/Endpoints/Resolve.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        switch ($parent->getObjectType()) {
+            case 'page_id':
+                return $this->notion->pages()->find($parent->getId());
+            case 'database_id':
+                return $this->notion->databases()->find($parent->getId());
+            case 'block_id':
+                return $this->notion->block($parent->getId())->retrieve();
+            case 'workspace':
+                return $parent;
+                // throw new HandlingException('A Notion Workspace cannot be resolved by the Notion API.');
```
[the remark this answers is not in the document]
Comment (johguentner) [reply src/Endpoints/Resolve.php:88]: yep, forgot it there, done :)

File: src/Builder/DatabaseBuilder.php (hunk 4 above)
Comment (johguentner) [reply src/Builder/DatabaseBuilder.php:222]: yep!

File: src/Builder/DatabaseBuilder.php (hunk 3 above)
Comment (johguentner) [reply src/Builder/DatabaseBuilder.php:170]: it's right there, isn't it ?

File: src/Builder/DatabaseBuilder.php (hunk 2 above)
Comment (johguentner) [reply src/Builder/DatabaseBuilder.php:46]: yeps, done!

File: src/Builder/DatabaseBuilder.php (hunk 1 above)
Comment (johguentner) [reply src/Builder/DatabaseBuilder.php:24]: yes, done!

**Pull request merged**